### ISSUE TITLE
kneighbors fail

### ISSUE THREAD

**Issue by eugene-yang** [FreeDiscovery/FreeDiscovery#126]
I encountered this problem when using kneighbors.

The version I'm using is b3997495b341c42e8946a4326db2570224b14c1c.

This problem only occurs when doing multiple iterations of training. No matter what initial seed set(training documents) was chosen, this error would appear after 43 or 45 iterations. 
The code is like this:
```python
fnames_all, X_all = fe.load(dataset_id)
...
from freediscovery.categorization import NearestNeighborRanker
mod = NearestNeighborRanker(n_jobs=1)
y_pred, _, md = mod.kneighbors(X_all)
```
The following is the error message. 
```
Traceback (most recent call last):
  File "run_trial.py", line 436, in <module>
    y_pred, _, md = mod.kneighbors(X_all)
  File "/lustre/home/ey120/Repositories/FreeDiscovery/freediscovery/categorization.py", line 186, in kneighbors
    D_n, idx_n_loc = self._mod_n.kneighbors(X)
  File "/home/ey120/local/anaconda3/envs/freediscovery-env/lib/python3.6/site-packages/sklearn/neighbors/base.py", line 353, in kneighbors
    n_jobs=n_jobs, squared=True)
  File "/home/ey120/local/anaconda3/envs/freediscovery-env/lib/python3.6/site-packages/sklearn/metrics/pairwise.py", line 1240, in pairwise_distances
    return _parallel_pairwise(X, Y, func, n_jobs, **kwds)
  File "/home/ey120/local/anaconda3/envs/freediscovery-env/lib/python3.6/site-packages/sklearn/metrics/pairwise.py", line 1083, in _parallel_pairwise
    return func(X, Y, **kwds)
  File "/home/ey120/local/anaconda3/envs/freediscovery-env/lib/python3.6/site-packages/sklearn/metrics/pairwise.py", line 245, in euclidean_distances
    distances = safe_sparse_dot(X, Y.T, dense_output=True)
  File "/home/ey120/local/anaconda3/envs/freediscovery-env/lib/python3.6/site-packages/sklearn/utils/extmath.py", line 186, in safe_sparse_dot
    ret = ret.toarray()
  File "/home/ey120/local/anaconda3/envs/freediscovery-env/lib/python3.6/site-packages/scipy/sparse/compressed.py", line 920, in toarray
    return self.tocoo(copy=False).toarray(order=order, out=out)
  File "/home/ey120/local/anaconda3/envs/freediscovery-env/lib/python3.6/site-packages/scipy/sparse/coo.py", line 258, in toarray
    B.ravel('A'), fortran)
ValueError: could not convert integer scalar
```

**Comment by rth** [comment]
@eugene-yang  Thanks for reporting this issue.

This error comes from scipy (see https://github.com/scipy/scipy/issues/7230) and happens if the NN distance matrix have more than 2**31 elements. Which means for a dataset of 700k documents, that the training set should be below ~2k documents (which would need ~60 GB RAM BTW). This is with FreeDiscovery from January https://github.com/FreeDiscovery/FreeDiscovery/commit/b3997495b341c42e8946a4326db2570224b14c1c.

In FreeDiscovery >= v0.9.0 this wouldn't happen since we [chunk the test set in NN to keep the RAM requirements to a few GB](https://github.com/FreeDiscovery/FreeDiscovery/blob/master/freediscovery/neighbors.py#L41). We should upgrade the experimental setup to the latest FreeDiscovery version..

**Comment by rth** [comment]
@eugene-yang I didn't see the broader problem initially, only a scipy bug with sparse arrays, but it only makes sense to use the Nearest Neighbor classifier in the LSI space because of the curse of dimensionality. So in this case `X_all` should have been dense and this issue wouldn't have occurred in the first place. 

In FreeDiscovery (or the experimental setup we are using), the use of LSI with Nearest Neighbors classifier is not explicitly enforced, but it probably should be to avoid this problem.

**Comment by eugene-yang** [comment]
@rth I agree with this hypothesis. I didn't use LSI for these testing. I will post further results of testing Nearest Neighbor with LSI couple days later.

**Comment by rth** [comment]
Closing as fixed in FreeDiscovery 1.0, please reopen if you still experience this issue.
